lib/ash/notifier/pub_sub/topic_generator.ex

```diff
@@ -80,25 +82,44 @@ defmodule Ash.Notifier.PubSub.TopicGenerator do
 
   Should return a list of strings representing the topics to publish to.
   An empty list means no topics will be published.
+  """
+  @callback generate_topics(notification :: Ash.Notifier.Notification.t(), opts :: keyword()) ::
+              String.t()
+              | [
+                  String.t()
+                ]
+
+  @doc """
+  Declare what relationships should be preloaded for efficient topic generation.
+
+  This optional callback allows topic generators to specify relationship
+  loading requirements. When implemented, the specified relationships will
+  be automatically loaded before the notification is created, eliminating
+  N+1 query problems in bulk operations.
+
+  ## Returns
+
+  Any load statement compatible with `Ash.load/2`:
+  - `:relationship_name` - Load a single relationship
+  - `[:rel1, :rel2]` - Load multiple relationships  
+  - `[author: :followers]` - Load nested relationships
+  - `[author: {:followers, query}]` - Load with custom query
 
   ## Examples
 
   ```elixir
-  def generate_topics(notification, _opts) do
-    case notification.data do
-      %{author_id: author_id} when not is_nil(author_id) ->
-        ["author_feed:\#{author_id}"]
-      _ ->
-        []
-    end
+  def required_loads do
+    [author: :followers]
+  end
+
+  def required_loads do
+    [:author, :tags, :comments]
   end
   ```
   """
-  @callback generate_topics(notification :: Ash.Notifier.Notification.t(), opts :: keyword()) ::
-              String.t()
-              | [
-                  String.t()
-                ]
+  @callback required_loads() :: term()
```

Comment:
This should be a callback on `Ash.Notifier` I think.

Reply:
Ah, nvm. This makes sense here. But my other comment about having a callback on the notifier itself to drive this is the thing.